### PR TITLE
Improve registration workflow

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | kinda
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | continuation of #13097
| License         | MIT                                                          |

I rebased the PR, but couldn't push it due to permissions issues. I pushed the rebased branch onto my fork, and created this replacement PR (ofc. authorship behaved :))

### CODE REVIEW

File: features/account/receiving_email_after_registration.feature [hunk 1]
```diff
@@ -9,11 +9,29 @@ Feature: Receiving a welcoming email after registration
...
     @ui @email @api
-    Scenario: Receiving a welcoming email after registration
+    Scenario: Receiving a welcoming email after registration when channel has disabled registration verification
+        Given on this channel account verification is not required
         When I register with email "ghastly@bespoke.com" and password "suitsarelife"
-        Then a welcoming email should have been sent to "ghastly@bespoke.com"
+        Then only one email should have been sent to "ghastly@bespoke.com"
+        And a welcoming email should have been sent to "ghastly@bespoke.com"
+
+    @ui @email
```
Comment (AdamKasp) [inline features/account/receiving_email_after_registration.feature]: Why this scenario is not cover by API?
we should add `@api` tag or `@no-api` tag (same in line 32)

File: features/account/receiving_email_after_registration.feature [hunk 2]
```diff
@@ -9,11 +9,29 @@ Feature: Receiving a welcoming email after registration
...
+    Scenario: Receiving an account verification email after registration when channel has enabled registration verification
+        Given on this channel account verification is required
+        When I register with email "ghastly@bespoke.com" and password "suitsarelife"
+        Then only one email should have been sent to "ghastly@bespoke.com"
+        And a verification email should have been sent to "ghastly@bespoke.com"
+        And a welcoming email should not have been sent to "ghastly@bespoke.com"
```
Comment (AdamKasp) [inline features/account/receiving_email_after_registration.feature]: it super minor but here we can (probably) use key word `but`
```suggestion
        And a verification email should have been sent to "ghastly@bespoke.com"
        But a welcoming email should not have been sent to "ghastly@bespoke.com"
```

File: features/account/registering.feature [hunk 3]
```diff
@@ -17,6 +18,7 @@ Feature: Account registration
         And I confirm this password
         And I register this account
         Then I should be notified that new account has been successfully created
+        And I should be on registration thank you page
```
Comment (AdamKasp) [inline features/account/registering.feature:21]: I know that we can not fix it quickly but this step has no sense in API context :(

File: src/Sylius/Behat/Context/Api/EmailContext.php [hunk 4]
```diff
@@ -169,6 +169,14 @@ public function recipientShouldReceiveNoEmails(string $recipient): void
         Assert::false($this->emailChecker->hasRecipient($recipient));
     }
 
+    /**
+     * @Then /^only one email should have been sent to "([^"]*)"$/
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/EmailContext.php]: do we need to use regex here?
IIRC by default we use:
```suggestion
     * @Then only one email should have been sent to :recipientEmail
```

Comment (AdamKasp) [reply src/Sylius/Behat/Context/Api/EmailContext.php]: commend valid alos to next 2 files

File: src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php [hunk 5]
```diff
@@ -251,6 +251,22 @@ public function iShouldBeSubscribedToTheNewsletter(): void
...
+    }
+
+    /**
+     * @Then /^I should be on my account dashboard/
+     */
+    public function iShouldBeOnMyAccountDashboard(): void
+    {
+        // Intentionally left blank
+    }
+
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php]: I feel we can merge it into single step

File: src/Sylius/Behat/Context/Ui/EmailContext.php [hunk 6]
```diff
@@ -166,11 +188,24 @@ public function recipientShouldReceiveNoEmails(string $recipient): void
...
+    {
+        Assert::eq($this->emailChecker->countMessagesTo($recipient), 1);
+    }
+
     private function assertEmailContainsMessageTo(string $message, string $recipient): void
     {
         Assert::true($this->emailChecker->hasMessageTo($message, $recipient));
     }
 
+    private function assertEmailDoesNotContainMessageTo(string $message, string $recipient): void
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Ui/EmailContext.php:204]:
```suggestion
    private function assertEmailDoesNotContainsMessageTo(string $message, string $recipient): void
```

File: src/Sylius/Behat/Page/Shop/Account/RegisterThankYouPageInterface.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Sylius\Behat\Page\Shop\Account;
+
+use FriendsOfBehat\PageObjectExtension\Page\SymfonyPageInterface;
+
+interface RegisterThankYouPageInterface extends SymfonyPageInterface
+{
+
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Account/RegisterThankYouPageInterface.php]:
```suggestion
```

File: src/Sylius/Behat/Resources/config/services/contexts/ui.xml [hunk 8]
```diff
@@ -417,7 +417,10 @@
             <argument type="service" id="sylius.behat.page.shop.order.thank_you" />
             <argument type="service" id="sylius.behat.page.shop.home" />
             <argument type="service" id="sylius.behat.page.shop.account.verify" />
+            <argument type="service" id="sylius.behat.page.shop.account.register.thank_you" />
+            <argument type="service" id="sylius.behat.page.shop.account.dashboard" />
             <argument type="service" id="sylius.behat.element.shop.account.register" />
+            <argument type="service" id="sylius.repository.customer" />
```
Comment (AdamKasp) [inline src/Sylius/Behat/Resources/config/services/contexts/ui.xml:426]: I am not sure, but what is it?

File: src/Sylius/Behat/Context/Ui/EmailContext.php (hunk 6 above)
Comment (jakubtobiasz) [reply src/Sylius/Behat/Context/Ui/EmailContext.php:204]: ✅ It doesn't contain
❌ It doesn't contain**s**

The function's name is correct :D.

File: src/Sylius/Behat/Resources/config/services/contexts/ui.xml (hunk 8 above)
Comment (jakubtobiasz) [reply src/Sylius/Behat/Resources/config/services/contexts/ui.xml:426]: CleanShot 2023-06-28 at 10 16 44
It is used to find a last registered user. This solution is proposed by an original author, maybe we can check whether a “Thank You” message is displayed as in other similar cases. I don't have a strong opinion 🤷🏻.

File: features/account/registering.feature (hunk 3 above)
Comment (GSadee) [reply features/account/registering.feature:21]: Maybe create a separate scenario for API context here (without this step and with some steps merged)?

File: features/checkout/registering_account_after_checkout.feature [hunk 9]
```diff
@@ -20,5 +21,19 @@ Feature: Registering a new account after checkout
         And I specify a password as "sylius"
         And I confirm this password
         And I register this account
-        And I verify my account using link sent to "john@example.com"
+        Then I should be on registration thank you page
+        When I verify my account using link sent to "john@example.com"
         Then I should be able to log in as "john@example.com" with "sylius" password
```
Comment (GSadee) [inline features/checkout/registering_account_after_checkout.feature]: We shouldn't mix `Then`, `When` and `Then` steps in a single scenario. Some steps should be changed or the logic should be split into two scenarios

File: features/checkout/registering_account_after_checkout.feature [hunk 10]
```diff
@@ -20,5 +21,19 @@ Feature: Registering a new account after checkout
...
+
+    @ui @no-api
+    Scenario: Registering a new account after checkout when channel has disabled registration verification
+        Given on this channel account verification is not required
+        Given I have product "PHP T-Shirt" in the cart
```
Comment (GSadee) [inline features/checkout/registering_account_after_checkout.feature]:
```suggestion
        And I have product "PHP T-Shirt" in the cart
```

File: src/Sylius/Behat/Context/Api/EmailContext.php [hunk 11]
```diff
@@ -169,11 +191,24 @@ public function recipientShouldReceiveNoEmails(string $recipient): void
         Assert::false($this->emailChecker->hasRecipient($recipient));
     }
 
+    /**
+     * @Then only one email should have been sent to :recipientEmail
+     */
+    public function onlyOneEmailShouldHaveBeenSentTo(string $recipient): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/EmailContext.php]:
```suggestion
     * @Then only one email should have been sent to :recipientEmail
     */
    public function onlyOneEmailShouldHaveBeenSentTo(string $recipientEmail): void
```

File: src/Sylius/Behat/Context/Ui/EmailContext.php [hunk 12]
```diff
@@ -166,11 +188,24 @@ public function recipientShouldReceiveNoEmails(string $recipient): void
         Assert::false($this->emailChecker->hasRecipient($recipient));
     }
 
+    /**
+     * @Then /^only one email should have been sent to "([^"]*)"$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/EmailContext.php]:
```suggestion
     * @Then only one email should have been sent to :recipient
```

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/RegistrationAfterCheckoutContext.php [hunk 13]
```diff
@@ -93,4 +99,21 @@ public function iShouldBeAbleToLogInAsWithPassword(string $email, string $passwo
...
+     * @Then I should be on registration thank you page
+     */
+    public function iShouldBeOnRegistrationThankYouPage(): void
+    {
+        $registeredCustomer = $this->customerRepository->findLatest(1)[0];
+        Assert::true($this->registerThankYouPage->isOpen(['id' => $registeredCustomer->getId()]));
+    }
+
+    /**
+     * @Then /^I should be on my account dashboard/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/RegistrationAfterCheckoutContext.php]:
```suggestion
     * @Then I should be on my account dashboard
```

File: src/Sylius/Behat/Context/Ui/Shop/RegistrationContext.php [hunk 14]
```diff
@@ -348,6 +352,23 @@ public function iShouldBeSubscribedToTheNewsletter(): void
         Assert::true($this->profileUpdatePage->isSubscribedToTheNewsletter());
     }
 
+    /**
+     * @Then /^I should be on registration thank you page$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/RegistrationContext.php]:
```suggestion
     * @Then I should be on registration thank you page
```

File: src/Sylius/Behat/Context/Ui/Shop/RegistrationContext.php [hunk 15]
```diff
@@ -348,6 +352,23 @@ public function iShouldBeSubscribedToTheNewsletter(): void
...
+     */
+    public function iShouldBeOnRegistrationThankYouPage(): void
+    {
+        $registeredCustomer = $this->customerRepository->findLatest(1)[0];
+        Assert::true($this->registerThankYouPage->isOpen(['id' => $registeredCustomer->getId()]));
+    }
+
+    /**
+     * @Then /^I should be on my account dashboard$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/RegistrationContext.php]:
```suggestion
     * @Then I should be on my account dashboard
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Account/VerifyCustomerAccountHandler.php [hunk 16]
```diff
@@ -44,6 +52,14 @@ public function __invoke(VerifyCustomerAccount $command): JsonResponse
         $user->setEmailVerificationToken(null);
         $user->enable();
 
+        $channel = $this->channelContext->getChannel();
+        $localeCode = $this->localeContext->getLocaleCode();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Account/VerifyCustomerAccountHandler.php]: IMO this handler should not use contexts to get these data, it should be provided by the command.

File: src/Sylius/Bundle/ShopBundle/Controller/RegistrationThankYouController.php [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function thankYouAction(): Response
+    {
+        /** @var ChannelInterface $channel */
+        $channel = $this->channelContext->getChannel();
+
+        if ($channel->isAccountVerificationRequired()) {
+            return new Response($this->twig->render('@SyliusShop/registerThankYou.html.twig'));
+        }
+        return new RedirectResponse($this->router->generate('sylius_shop_account_dashboard'));
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Controller/RegistrationThankYouController.php:31]:
```suggestion

        return new RedirectResponse($this->router->generate('sylius_shop_account_dashboard'));
```

File: src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml [hunk 18]
```diff
@@ -846,6 +846,7 @@ sylius:
         textarea: 'Textarea'
         thank_you: 'Thank you!'
         thank_you_for_your_order: 'Thank you for your order'
+        thank_you_for_register: 'Thank you for your registration'
```
Comment (GSadee) [inline src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml]:
```suggestion
        thank_you_for_register: 'Thank you for your registration'
        thank_you_for_your_order: 'Thank you for your order'
```

File: src/Sylius/Behat/Context/Api/EmailContext.php (hunk 11 above)
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/EmailContext.php]:
```suggestion
     * @Then only one email should have been sent to :recipient
     */
    public function onlyOneEmailShouldHaveBeenSentTo(string $recipient): void
```
I'd go the opposite way to be in line with the other methods

File: src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml (hunk 18 above)
Comment (NoResponseMate) [inline src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml]:
```suggestion
        thank_you_for_your_registration: 'Thank you for your registration'
```
A bit more consistency

File: src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php [hunk 19]
```diff
@@ -251,6 +251,14 @@ public function iShouldBeSubscribedToTheNewsletter(): void
         Assert::true($this->responseChecker->getResponseContent($response)['subscribedToNewsletter']);
     }
 
+    /**
+     * @Then I should be on my account dashboard
+     * @Then I should be on registration thank you page
+     */
+    public function intentionallyLeftBlank(): void
```
Comment (TheMilek) [inline src/Sylius/Behat/Context/Api/Shop/RegistrationContext.php:258]: I would stick to the convention of naming methods like we do in other steps, so make 2 blanks or one universal like
```suggestion
    public function iShouldBeOn(): void
```
but still it's just a suggestion ✌🏽

**Review by TheMilek — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Comment by TheMilek** [comment]
Thanks, Jacob! :tada:

**Pull request merged**